streams/src/test/java/org/apache/kafka/streams/integration/EosBetaUpgradeIntegrationTest.java

```diff
@@ -0,0 +1,691 @@
+        // 2.  write 10 records per input topic partition and verify that the result was committed
+        // 3.  write 5 records per input topic partition to get pending transactions (verified via "read_uncommitted" mode)
+        //      - both pending transactions are based on task producers
+        // 4.  stop/crash the first client, wait until rebalance stabilizes:
+        //      - stop case:
+        //        * verify that the stopped client did commit its pending transaction during shutdown
+        //        * the second client will still have an pending transaction
+        //      - crash case:
+        //        * the second client will have two pending transactions
+        // 5.  restart the first client with eos-beta enabled and wait until rebalance stabilizes
+        //       - crash only: the rebalance should result in a commit
+        // 6.  write 5 record per input topic partition and verify that the result was committed
+        // 7.  write 5 records per input topic partition to get pending transactions (verified via "read_uncommitted" mode)
+        //      - one transaction is base on a task producer; the other transaction is based on a thread producer
+        // 8.  stop/crash the second client and wait until rebalance stabilizes:
+        //      - stop only: verify that the stopped client did commit its pending transaction during shutdown
+        //      - the first client will still have an pending transaction
+        // 9.  restart the second client with eos-beta enabled and wait until rebalance stabilizes
+        // 10. write 5 record per input topic partition and verify that the result was committed
+
+        try (final KafkaStreams streams1Alpha = getKafkaStreams("appDir1", StreamsConfig.EXACTLY_ONCE);
+            final KafkaStreams streams1Beta = getKafkaStreams("appDir1", StreamsConfig.EXACTLY_ONCE_BETA);
+            final KafkaStreams streams2Alpha = getKafkaStreams("appDir2", StreamsConfig.EXACTLY_ONCE);
+            final KafkaStreams streams2Beta = getKafkaStreams("appDir2", StreamsConfig.EXACTLY_ONCE_BETA)) {
+
+            final List<KeyValue<KafkaStreams.State, KafkaStreams.State>> stateTransitions1 = new LinkedList<>();
+            streams1Alpha.setStateListener((newState, oldState) -> stateTransitions1.add(KeyValue.pair(oldState, newState)));
+            streams1Beta.setStateListener((newState, oldState) -> stateTransitions1.add(KeyValue.pair(oldState, newState)));
+
+            final List<KeyValue<KafkaStreams.State, KafkaStreams.State>> stateTransitions2 = new LinkedList<>();
+            streams2Alpha.setStateListener((newState, oldState) -> stateTransitions2.add(KeyValue.pair(oldState, newState)));
+            streams2Beta.setStateListener((newState, oldState) -> stateTransitions2.add(KeyValue.pair(oldState, newState)));
+
+            // phase 1:
+            final AtomicBoolean errorInjectedClient1 = new AtomicBoolean(false);
+            errorInjected = errorInjectedClient1;
```

Comment:
Seems we can just keep errorInjected to null in phase 1 since we do not need to inject failures?

Reply:
This is a little tricky... We get a handle on the `AtomicBoolean` within the `Processor#init()` method. In phase 1 we create the `Processor` and thus we need to prepare the error injection at this point already. In the later phases, `Processor#init()` is not called for the processors hosted by client-1 any more.

Does this make sense?

Reply:
Got it, makes sense.